src/opportunity_detector.py

```diff
@@ -0,0 +1,165 @@
+import networkx as nx
+from typing import List, Tuple, Optional
+
+# Assuming ShortTicker and Symbol classes are defined elsewhere
+# Here's a basic implementation for completeness
+class Symbol:
+    def __init__(self, base, quote):
+        self.base = base
+        self.quote = quote
+
+class ShortTicker:
+    def __init__(self, symbol: Symbol, last_price: float, reversed: bool = False):
+        self.symbol = symbol
+        self.last_price = last_price
+        self.reversed = reversed
+
+    def __repr__(self):
+        direction = f"{self.symbol.base}/{self.symbol.quote}"
+        if self.reversed:
+            direction += " (Reversed)"
+        return f"{direction} @ {self.last_price}"
+
+
+# A simple cycle, or elementary circuit, is a closed path where no node appears twice.
+# In a directed graph, two simple cycles are distinct if they are not cyclic permutations of each other. 
+# In an undirected graph, two simple cycles are distinct if they are not cyclic permutations of each other nor of the other’s reversal.
+
+# Optionally, the cycles are bounded in length. In the unbounded case, we use a nonrecursive, iterator/generator version of Johnson’s algorithm.
+# In the bounded case, we use a version of the algorithm of Gupta and Suzumura. There may be better algorithms for some cases.
+
+# The algorithms of Johnson, and Gupta and Suzumura, are enhanced by some well-known preprocessing techniques. 
+# When G is directed, we restrict our attention to strongly connected components of G, generate all simple cycles
+# containing a certain node, remove that node, and further decompose the remainder into strongly connected components.
+# When G is undirected, we restrict our attention to biconnected components, generate all simple cycles containing a particular edge,
+# remove that edge, and further decompose the remainder into biconnected components.
+def get_triangular_opportunity(tickers: List['ShortTicker']) -> Tuple[List['ShortTicker'], float]:
+    # Build a directed graph of currencies
+    graph = nx.DiGraph()
+
+    for ticker in tickers:
+        if ticker.symbol is not None:
+            # Add forward edge
+            graph.add_edge(ticker.symbol.base, ticker.symbol.quote, ticker=ticker)
+            # Add reverse edge with inverted price
+            reversed_ticker = ShortTicker(Symbol(ticker.symbol.quote, ticker.symbol.base),
+                                          1 / ticker.last_price, reversed=True)
+            graph.add_edge(ticker.symbol.quote, ticker.symbol.base, ticker=reversed_ticker)
+
+    best_profit = 0
+    best_triplet = None
+
+    # Find all cycles in the graph (specifically triplets)
+    for cycle in nx.simple_cycles(graph):
+        if len(cycle) != 3:
```

Comment:
@samdevo If you look at the original code from their triangular arbitrage opportunity detection, the time complexity is the same as the algorithm is the same.